tutorials/cloud-storage-infinite-compose/index.md

```diff
@@ -0,0 +1,305 @@
+
+
+```python
+def compose_and_cleanup(blob: storage.Blob, chunk: List[storage.Blob],
+                        client: storage.Client, executor: Executor):
+    """Compose a blob and clean up its components. Cleanup tasks will be
+    scheduled in the provided executor and the composed blob immediately
+    returned.
+
+    Args:
+        blob (storage.Blob): The blob to be composed.
+        chunk (List[storage.Blob]): The component blobs.
+        client (storage.Client): A Cloud Storage client.
+        executor (Executor): An executor in which to schedule cleanup tasks.
+
+    Returns:
+        storage.Blob: The composed blob.
+    """
+    # wait on results if the chunk has any futures
+    chunk = ensure_results(chunk)
+    blob.compose(chunk, client=client)
+    # cleanup components, no longer need them
+    delete_objects_concurrent(chunk, executor, client)
+    return blob
+```
+
+Something of a stylistic choice, the `ensure_results` function serves to let us accept either a list of storage.Blob (which we have from our initial list) or a list of Future[storage.Blob] (which we have from intermediate steps), or a mix of both, using the same code. This is what makes our compose_and_cleanup function block until its components are all complete.
+
+```python
+def ensure_results(maybe_futures: List[Any]) -> List[Any]:
+    """Pass in a list that may contain a Future(s), and if so, wait for
+    the result of the Future; for all other types in the list,
+    simply append the value.
+
+    Args:
+        maybe_futures (List[Any]): A list which may contain Futures.
+
+    Returns:
+        List[Any]: A list with the values passed in, or Future.result() values.
+    """
+    results = []
+    for mf in maybe_futures:
+        if isinstance(mf, Future):
+            results.append(mf.result())
+        else:
+            results.append(mf)
+    return results
+```
+
+Finally, this function simply gives us a very long sequence of hexadecimal digits to use as identifiers for our intermediate accumulators.
+
+```python
+def generate_hex_sequence() -> Iterable[str]:
+    """Generate an indefinite sequence of hexadecimal integers.
+
+    Yields:
+        Iterator[Iterable[str]]: The sequence of hex digits, as strings.
+    """
+    for i in count(0):
+        yield hex(i)[2:]
+```
+
+This approach trades off some complexity for a great increase in speed. In [this example command line tool](https://github.com/domZippilli/gcsfast/blob/master/gcsfast/cli/upload_standard.py), the technique is used to upload objects using parallel slices.
+
+
+## Conclusion
+
+Hopefully, this article has shown you how you can use the Cloud Storage compose function to create single objects from very large collections of other objects. It's a great thing to be able to do. I want to leave you with a few caveats:
+
+*   Remember that while there's no limit to the number of components, the 5TB object size limit does still apply. 
```

Comment:
Although the number is very high.
Do you feel putting a number to [max number of components](https://cloud.google.com/storage/quotas#:~:text=Up%20to%2032%20objects%20can,composite%20object%20saturates%20at%202%2C147%2C483%2C647) (2,147,483,647) from a metadata perspective makes sense?

Reply:
I linked to that page on [line 28](https://github.com/GoogleCloudPlatform/community/pull/1531/files#diff-7c838efdb3715fb5f68b4a0c2bda8d2124ec27d4a2743ead2940babac0a5ad6eR28) and explained the integer saturation, so I think there's lots of information about this already.

Reply:
I think that we're well covered on this.